### PR TITLE
support search in SICP JS 

### PR DESCRIPTION
**Pull request by yiwen101**
I add search functionality to the current NavigationBarSicp component.

need to work together with my modification for SICP backend which generate searchDataFile
run yarn json to generate searchjson Data before running yarn tryjson in the SICP backend, and run yarn start at the frontend

### CODE REVIEW

**Review by martin-henz — APPROVED** [review]
Nice from-scratch implementation of trie.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 1]
```diff
@@ -1,8 +1,12 @@
-import { Alignment, Drawer, Navbar, NavbarGroup, Position } from '@blueprintjs/core';
+import { Alignment, Drawer, InputGroup, Navbar, NavbarGroup, Position } from '@blueprintjs/core';
 import { IconNames } from '@blueprintjs/icons';
+import { memoize } from 'lodash';
 import * as React from 'react';
+//import React, {useEffect} from 'react';
+//import Highlighter from 'react-highlight-words';
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:6]: Commented code should be removed.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 2]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+        node = node[char];
+      } else {
+        return [];
+      }
+    }
+    if (node['value']) {
+      return node['value'];
+    }
+    return [];
+  }
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:143]: There's a lot of duplicated code here. There should just be one function, with the trie being queried passed in as an argument instead of relying on closure.

Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:143]: I've fixed this in #2471. Do you mind taking a look?

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 3]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+
+  function autoComplete(str: string, limit: number, jsonData: any) {
+    if (str.length === 0) {
+      return [];
+    }
+    str = str.toLowerCase();
+    const trie = jsonData;
+
+    function next(node: any, i: number) {
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:152]: If we are using `any` like this, it's ideally only temporary, but more importantly, as it presents a lot of potential type safety issues, it should be aliased.

```ts
type TrieNode = any;
// Then, wherever `any` refers to a node in the trie...
function next(node: TrieNode, i: number) {
  // ...
}
```

This will not only greatly add readability and prevent bugs, but also developer experience (from completions, etc).

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 4]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+      return {
+        title: textbook[array[0]][array[1]][array[2]],
+        url: SearchUrl + array[0] + array[1]
+      };
+    }
+    const SearchUrl = '..';
+    let words = str.toLowerCase().split(' ');
+    words = words.filter(word => word !== '');
+    if (searchQuery.length === 0 || words.length === 0) {
+      setQueryResult([]);
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:233]: Why is it set as an empty array here, but set as `[{ title: 'no result found', url: '' }]` in the initial state (when there are also no results)?

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 5]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+      return;
+    }
+    let globalAns = queryTextTrie(words[0]).map((array: any) => lineMap(array));
+
+    globalAns = globalAns.filter((obj: any) => obj.title.toLowerCase().includes(str.toLowerCase()));
+    return globalAns;
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:239]: Let us prefer to use `const`. The mutation here is unnecessary:

```suggestion
    const globalAns = queryTextTrie(words[0]).map((array: any) => lineMap(array));
    return globalAns.filter((obj: any) => obj.title.toLowerCase().includes(str.toLowerCase()));
```

Also note the use of the unaliased `any` type (not addressed in the suggestion above).

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 1 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:6]: I've fixed this in #2471.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 5 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:239]: I've fixed this in #2471.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 6]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+
+    function recur(node: any, path: string) {
+      if (node == null || limit < 0) {
+        return;
+      }
+      if (node.value) {
+        ans.push(path);
+        limit--;
+      }
+      for (let i = 0; i < 27; i++) {
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:210]: What is 27? Is this a magic number?

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 7]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+      return [];
+    }
+    str = str.toLowerCase();
+    const trie = jsonData;
+
+    function next(node: any, i: number) {
+      if (i === 26) return node[' '];
+      else if (i >= 0 && i <= 25) {
+        return node[String.fromCharCode(i + 97)];
+      }
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:156]: Where are all these magic numbers coming from?

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 8]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+      if (array == null || array[0] == null || array[1] == null || array[2] == null) {
+        return { title: '', url: '' };
+      }
+      return {
+        title: textbook[array[0]][array[1]][array[2]],
+        url: SearchUrl + array[0] + array[1]
+      };
+    }
+    const SearchUrl = '..';
+    let words = str.toLowerCase().split(' ');
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:230]: What does this function do? The naming is unclear and it is not helped by the following:
* everything being typed as `any`
* non-descriptive variables names (e.g. `array` – what array?, `str` parameter – what string is this??)
* the code having zero comments

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 9]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+    function lineMap(array: any[]) {
+      if (array == null || array[0] == null || array[1] == null || array[2] == null) {
+        return { title: '', url: '' };
+      }
+      return {
+        title: textbook[array[0]][array[1]][array[2]],
+        url: SearchUrl + array[0] + array[1]
+      };
+    }
+    const SearchUrl = '..';
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:229]: What does this `'..'` mean?

**Review by RichDom2185 — COMMENTED** [review]
Quite a lot of severe type-safety and readability issues. I hope my comments aren't too mean. I can help with the fixes; can you respond to the comments for clarification?

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 1 above)
Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:6]: Thank you for your amazing efforts and helps!

Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:6]: I feel very sorry for the trouble caused by my incompetency; Thank you for your fixes in 2471; will definitely pay more attention to code quality in terms of type safety and readability.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 6 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:210]: Similarly, what does the `voidSearch` function do? The readability is slightly hampered by the functions being called out-of-order they're defined in.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 7 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:156]: No worries, I'll fix it. Thanks for the clarification!

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 8 above)
Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:230]: I intend to have the array as the returned array of "queryTrie", with information of the queried result stored in different index of it;

lineMap takes the result of the queryTrie, and transformed it to a format that is suitable for generating content that correspond to a line in the search result displayed to the user in web

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 6 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:210]: In that case, what does `queryTrie` do?

Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:210]: for index search, since all the index are in the IndexTrie, we can just use queryTire index;

But for user search, since the textTrie only contains the occurrences of individual words rather than sentences, for example, "instruction", but not "follow instructions", I need to use voidSearch instead

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 9 above)
Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:229]: Sorry, I could not make sense of it either

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 10]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+      'x',
+      'y',
+      'z',
+      ' '
+    ];
+
+    function toEnd(query: any) {
+      let node = trie;
+      for (let i = 0; i < query.length; i++) {
+        const char = query[i];
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:192]: This can be replaced with a for-of loop.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 11]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+      .toLowerCase()
+      .split(' ')
+      .filter(word => word !== '');
+    if (words.length < 2) {
+      return autoComplete(str, limit, jsonData);
+    }
+    let pre = words[0];
+    for (let i = 1; i < words.length - 1; i++) {
+      pre += ' ' + words[i];
+    }
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:253]: String concatenation using `+=` is extremely slow. The `.join()` method can be used instead, since this is an array.

Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:253]: Fixed in #2471.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 10 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:192]: Fixed in #2471.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx [hunk 12]
```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
...
+    if (preResults.length < 1) {
+      return [];
+    }
+    let lastwords = autoComplete(lastWord, 3000, jsonData);
+    console.log(lastwords);
+    lastwords = lastwords.filter(
+      (word: string) =>
+        preResults.filter((obj: any) => obj.title.toLowerCase().includes(pre + ' ' + word)).length >
+        0
+    );
```
Comment (RichDom2185) [inline src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:271]: What is this code chunk trying to do? `.includes()` gives either `true` or `false`. What's the `.length > 0` for?

Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:271]: These are the types of issues that will be caused by having `any` for everything; it's even worse than writing in JavaScript as writing `any` disables all checks by your editor too.

Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:271]: Did you mean to just stop after `.includes()`? Can you explain how this entire function works, step by step, in plain English?

**Review by RichDom2185 — COMMENTED** [review]
Some more comments.

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 12 above)
Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:271]: sorry for bringing this up so late; Maybe you could skip deep "cleaning up" of my code, since many of them are meant to be abolished; I have extended it into CP3108B, and will keep on working on it next semester to properly clean it up

Comment (yiwen101) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:271]: As for this function, it goes like: it firstly separate the content that in the input line into two parts, parts that are completed and parts that are not. for example, "As for this fu" will be separated into ""As for this" and "fu". 


Then, "voidSearch" search for all the occurrences of the phrase "As for this" in the textbook coupled with the complete sentence they appear in the textbook. let us say there are only two results "As for this function" "as for this example".

Then, I use the autocomplete to generate guesses for the last word, for example, "fun", "function" "fuel"......

then I check whether they appear in the sentences and filter out absent ones. For example, only "function" appear in the  "As for this function" "as for this example".

so for my sentenceComplete, everyTime it can only produce the guess for the next one word (shrug)

File: src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx (hunk 9 above)
Comment (RichDom2185) [reply src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx:229]: Nevermind, this seems to be a path traversal.
